### PR TITLE
ENH: Added support for int-like strings and numpy dtypes

### LINKED ISSUE #1026: CRS.from_user_input does not accept numpy-integers
#### Code Sample, a copy-pastable example if possible

```python
# Your code here
from pyproj import CRS
import numpy as np

crs_input = 4326
crs_numpy_input = np.array([4326])[0]


crs1 = CRS.from_user_input(crs_input )

crs2 = CRS.from_user_input(crs_numpy_input)
```

```python
>>> CRSError: Invalid CRS input: 4326
```

#### Problem description

The (really nice) `CRS.from_user_input` function does not accept numpy-integers...
I guess it should be an easy fix to include `np.integer` as possible dtype.

#### Expected Output

I'd expect numpy-integers to be treated similar to python integers.

#### Environment Information

 - Output from: `pyproj -v`

pyproj info:
    pyproj: 3.2.1
    PROJ: 8.2.0

System:
    python: 3.7.12 | packaged by conda-forge | (default, Oct 26 2021, 05:37:49) [MSC v.1916 64 bit (AMD64)]
executable: D:\miniconda\envs\...\python.exe
   machine: Windows-10-10.0.17763-SP0

Python deps:
   certifi: 2021.10.08
       pip: 22.0.2
setuptools: 59.8.0
    Cython: None

### LINKED ISSUE #1035: ENH: Handle case when the SRID is just an integer, but defined as a string (e.g. "2972" instead of 2972)
In my current use case, I could receive the CRS definition simply as an integer (e.g. `2972`), or as a string, for cases when it is an ESRI-defined CRS (`e.g. "ESRI:102023"`). So I have a test suite with a CSV that contains, among other things, a column with this input. But since this column sometimes contains strings, everything gets downcasted to a string. It's not too hard to just check if the string is actually a number on my end (`.isnumeric()`), but it would be nice if pyproj could just detect this implicitly, because I think it is rather foolproof.

Around line 303 in `crs.py`, the initializer checks if the input value is a string or not, and if it is a string, handles it appropriately.

https://github.com/pyproj4/pyproj/blob/a8d1c0d64fa12b7a5081c40c1c94ff69b2c455d8/pyproj/crs/crs.py#L303

I suggest either adding a condition just before is checking `isinstance(projparams, str) and projparams.isnumeric()`, and casting it to an integer then and there. That way, `pyproj.crs.CRS(2972)` and `pyproj.crs.CRS("2972")` would both work.

I can add this if you approve.

### PR DESCRIPTION
**Pull request by iboates**
- [x] Closes #1026
 - [x] Closes #1035
 - [x] Tests added
 - [x] Fully documented, including `history.rst` for all changes and `api/*.rst` for new API

I don't think `api/*` needs updating but maybe I am wrong.

### CODE REVIEW

**Comment by snowman2** [comment]
Thoughts about instead adding a method called `_is_epsg_code` to check for numpy scalar, integer, or integer string and passing it into `from_epsg`?

`CRS.from_epsg` should be able to handle them if they are detected properly:
```python
>>> from pyproj import CRS
>>> import numpy
>>> CRS.from_epsg(numpy.array(4326))
<Geographic 2D CRS: EPSG:4326>
Name: WGS 84
Axis Info [ellipsoidal]:
- Lat[north]: Geodetic latitude (degree)
- Lon[east]: Geodetic longitude (degree)
Area of Use:
- name: World.
- bounds: (-180.0, -90.0, 180.0, 90.0)
Datum: World Geodetic System 1984 ensemble
- Ellipsoid: WGS 84
- Prime Meridian: Greenwich
>>> CRS.from_epsg("4326")
<Geographic 2D CRS: EPSG:4326>
Name: WGS 84
Axis Info [ellipsoidal]:
- Lat[north]: Geodetic latitude (degree)
- Lon[east]: Geodetic longitude (degree)
Area of Use:
- name: World.
- bounds: (-180.0, -90.0, 180.0, 90.0)
Datum: World Geodetic System 1984 ensemble
- Ellipsoid: WGS 84
- Prime Meridian: Greenwich
```

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -6,7 +6,7 @@ repos:
         -   id: end-of-file-fixer
         -   id: trailing-whitespace
     -   repo: https://github.com/psf/black
-        rev: 22.1.0
+        rev: 22.3.0
```
Comment (snowman2) [inline .pre-commit-config.yaml:9]: Just handled this in #1052. Mind reverting the pre-commit changes here?

Comment (iboates) [reply .pre-commit-config.yaml]: Done

**Comment by iboates** [comment]
@snowman2 Indeed a `_from_epsh_code` function is cleaner. But I am not sure what to specify as its parameter type, since it could be a numpy dtype, so wouldn't `object` be the only choice? and therefore make it irrelevant?

File: .pre-commit-config.yaml (hunk 1 above)
Comment (snowman2) [reply .pre-commit-config.yaml]: Thanks :+1:

**Comment by snowman2** [comment]
D

> @snowman2 Indeed a `_from_epsg_code` function is cleaner. But I am not sure what to specify as its parameter type, since it could be a numpy dtype, so wouldn't `object` be the only choice? and therefore make it irrelevant?

I like the current implementation, it looks good :+1:.

I think this would be sufficient:
```python
def _is_epsg_code(auth_code: Any) -> bool:
```

File: README.md [hunk 2]
```diff
@@ -99,6 +99,7 @@ Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/d
...
     <td align="center"><a href="https://github.com/martinfleis"><img src="https://avatars.githubusercontent.com/u/36797143?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Martin Fleischmann</b></sub></a><br /><a href="https://github.com/pyproj4/pyproj/issues?q=author%3Amartinfleis" title="Bug reports">🐛</a> <a href="https://github.com/pyproj4/pyproj/commits?author=martinfleis" title="Code">💻</a> <a href="https://github.com/pyproj4/pyproj/commits?author=martinfleis" title="Tests">⚠️</a></td>
+    <td align="center"><a href="https://github.com/iboates"><img src="https://avatars.githubusercontent.com/u/13814358?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Isaac Boates</b></sub></a><br /><a href="https://github.com/pyproj4/pyproj/issues?q=author%3Aiboates" title="Bug reports">🐛</a> <a href="https://github.com/pyproj4/pyproj/commits?author=iboates" title="Code">💻</a> <a href="https://github.com/pyproj4/pyproj/commits?author=iboates" title="Tests">⚠️</a></td>
```
Comment (snowman2) [inline README.md:102]: Mind removing this? It will make adding you easier after this is merged in (I have some tweaks I have to do).

Comment (iboates) [reply README.md]: Done

Comment (snowman2) [reply README.md]: Thanks :+1:

**Comment by snowman2** [comment]
@iboates would you like to squash your commits or would you mind if I did it on merge?

**Comment by iboates** [comment]
You can do it on merge it's fine

**Comment by iboates** [comment]
Realized that there was an `elif` after a `return` and the linter didnt like that

**Comment by snowman2** [comment]
Thanks @iboates :+1:

**Pull request merged**